msm/src/ffa/interpreter.rs

```diff
@@ -0,0 +1,110 @@
+use crate::{ffa::columns::FFAColumnIndexer, Ff1, N_LIMBS};
+use ark_ff::{PrimeField, Zero};
+use num_bigint::BigUint;
+use o1_utils::{field_helpers::FieldHelpers, foreign_field::ForeignElement};
+
+pub trait FFAInterpreterEnv<F: PrimeField> {
+    type Position;
+
+    type Variable: Clone
+        + std::ops::Add<Self::Variable, Output = Self::Variable>
+        + std::ops::Sub<Self::Variable, Output = Self::Variable>
+        + std::ops::Mul<Self::Variable, Output = Self::Variable>
+        + std::fmt::Debug;
+
+    fn empty() -> Self;
+
+    fn assert_zero(&mut self, cst: Self::Variable);
+
+    fn copy(&mut self, x: &Self::Variable, position: Self::Position) -> Self::Variable;
+
+    fn constant(value: F) -> Self::Variable;
+
+    fn column_pos(ix: FFAColumnIndexer) -> Self::Position;
+
+    fn read_column(&self, ix: FFAColumnIndexer) -> Self::Variable;
+
+    /// In constraint environment does nothing (?). In witness environment progresses to the next row.
+    fn next_row(&mut self);
+}
+
+// TODO use more foreign_field.rs with from/to bigint conversion
+fn limb_decompose<F: PrimeField>(input: &Ff1) -> [F; N_LIMBS] {
+    let input_bi: BigUint = FieldHelpers::to_biguint(input);
+    let ff_el: ForeignElement<F, N_LIMBS> = ForeignElement::from_biguint(input_bi);
+    ff_el.limbs
+}
+
+/// Reads values from limbs A and B, returns resulting value in C.
```

Comment:
I suppose it is not the appropriate place for this?

Reply:
Why isn't it? I mean the comment is not descriptive I agree :) But it's technically correct? I wanted to say what's the interface for this, that it's a constraint (thus function name) + it returns some values.

Reply:
I meant the function.